### ISSUE TITLE
Causal mask

### ISSUE THREAD

**Issue by pfeatherstone** [karpathy/nanoGPT#170]
Here: 
https://github.com/karpathy/nanoGPT/blob/ae3a8d5fdd3ddb8b13fab182723476523961e3ab/model.py#L57
you use `tril()`. Shouldn't it be `triu()` ?
You want the upper triangular attention matrix to be filled with -inf in order to get left attention right?

**Comment by davesque** [comment]
Using `tril` makes the lower triangular portion of `bias` contain all ones and the upper triangular portion off the diagonal contain all zeroes.  The upper triangular part is what we want filled with `float("-inf")` which is done on this line:
https://github.com/karpathy/nanoGPT/blob/ae3a8d5fdd3ddb8b13fab182723476523961e3ab/model.py#L76

Here's an example:
```python
In [1]: import torch

In [2]: tril = torch.tril(torch.ones(5, 5))

In [3]: tril
Out[3]:
tensor([[1., 0., 0., 0., 0.],
        [1., 1., 0., 0., 0.],
        [1., 1., 1., 0., 0.],
        [1., 1., 1., 1., 0.],
        [1., 1., 1., 1., 1.]])

In [4]: torch.randn(5, 5).masked_fill(tril == 0, float("-inf"))
Out[4]:
tensor([[ 1.7619,    -inf,    -inf,    -inf,    -inf],
        [-0.6219,  2.0269,    -inf,    -inf,    -inf],
        [ 0.2805, -1.0444,  1.1340,    -inf,    -inf],
        [-0.1109, -0.1044, -0.0911, -0.8428,    -inf],
        [ 0.4165, -0.5470,  1.1400, -0.0615, -0.6691]])
```

**Comment by pfeatherstone** [comment]
I missed the == 0.
If you used bias directly it would be triu()

**Comment by davesque** [comment]
Yep, true enough.  For that matter, you could also just do this:
```python
In [1]: import torch

In [2]: triu = torch.triu(torch.full((5, 5), float("-inf")), diagonal=1)

In [3]: triu
Out[3]:
tensor([[0., -inf, -inf, -inf, -inf],
        [0., 0., -inf, -inf, -inf],
        [0., 0., 0., -inf, -inf],
        [0., 0., 0., 0., -inf],
        [0., 0., 0., 0., 0.]])

In [4]: torch.randn(5, 5) + triu
Out[4]:
tensor([[ 0.0168,    -inf,    -inf,    -inf,    -inf],
        [ 1.3091, -0.7494,    -inf,    -inf,    -inf],
        [-2.0704, -1.1868, -1.4580,    -inf,    -inf],
        [ 1.1039,  0.9644,  0.3277, -2.0878,    -inf],
        [ 1.0196,  0.0292, -0.2291, -0.7627,  1.1352]])
```
Probably just comes down to personal preference and whichever one happens to be more performant.

**Comment by DarrenCook** [comment]
> `torch.triu(torch.full((5, 5), float("-inf")), diagonal=1)`

I like this: it is the most compact, and (IMHO) also more descriptive/readable than the alternatives. But I'd love to know if it is slower.

In my own code I've been creating it in what is the equivalent of class `GPT`'s `forward()`, using `T`, then pass it in for self-attention to use. Less memory usage (trivial difference), and it saves having to register a buffer for it. I wonder if it is slower or quicker overall though.

BTW, I noticed another very small optimization to the mask creation was suggested at https://github.com/karpathy/nanoGPT/issues/42#issuecomment-1379901986

**Comment by jeanmonet** [comment]
At the following line, shouldn't `attn_mask` be `self.bias`?

https://github.com/karpathy/nanoGPT/blob/a82b33b525ca9855d705656387698e13eb8e8d4b/model.py#L72

**Comment by rubenohana** [comment]
@jeanmonet No, according to the Pytorch documentation (https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html), because `is_causal = True`, it sets a causal attention mask, but would yield an error if both `attn_mask` and `is_causal` are both set. It is also shown in the `CausalSelfAttention` class of this page https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html, where `attn_mask=None`.
